src/test/java/edu/harvard/iq/dataverse/util/cache/CacheFactoryBeanTest.java

```diff
@@ -0,0 +1,360 @@
+            "        \"PublishDatasetCommand\"\n" +
+            "      ]\n" +
+            "    }\n" +
+            "  ]\n" +
+            "}";
+
+    @BeforeEach
+    public void init() throws IOException {
+        // Reuse cache and config for all tests
+        if (cache == null) {
+            mockedSystemConfig = mock(SystemConfig.class);
+            doReturn(settingDefaultCapacity).when(mockedSystemConfig).getRateLimitingDefaultCapacityTiers();
+            doReturn(settingJson).when(mockedSystemConfig).getRateLimitsJson();
+            cache = new CacheFactoryBean();
+            cache.systemConfig = mockedSystemConfig;
+            if (cache.rateLimitCache == null) {
+                cache.rateLimitCache = new TestCache(getConfig());
+            }
+
+            // Clear the static data, so it can be reloaded with the new mocked data
+            RateLimitUtil.rateLimitMap.clear();
+            RateLimitUtil.rateLimits.clear();
+        }
+
+        // Reset to default auth user
+        authUser.setRateLimitTier(1);
+        authUser.setSuperuser(false);
+        authUser.setUserIdentifier("authUser");
+
+        // Create a unique action for each test
+        action = "cmd-" + UUID.randomUUID();
+    }
+
+    @AfterAll
+    public static void cleanup() {
+        Hazelcast.shutdownAll();
+    }
+    @Test
+    public void testGuestUserGettingRateLimited() {
+        boolean rateLimited = false;
+        int cnt = 0;
+        for (; cnt <100; cnt++) {
+            rateLimited = !cache.checkRate(guestUser, action);
+            if (rateLimited) {
+                break;
+            }
+        }
+        String key = RateLimitUtil.generateCacheKey(guestUser, action);
+        assertTrue(cache.rateLimitCache.containsKey(key));
+        assertTrue(rateLimited && cnt > 1 && cnt <= 30, "rateLimited:"+rateLimited + " cnt:"+cnt);
+    }
+
+    @Test
+    public void testAdminUserExemptFromGettingRateLimited() {
+        authUser.setSuperuser(true);
+        authUser.setUserIdentifier("admin");
+        boolean rateLimited = false;
+        int cnt = 0;
+        for (; cnt <100; cnt++) {
+            rateLimited = !cache.checkRate(authUser, action);
+            if (rateLimited) {
+                break;
+            }
+        }
+        assertTrue(!rateLimited && cnt >= 99, "rateLimited:"+rateLimited + " cnt:"+cnt);
+    }
+
+    @Test
+    public void testAuthenticatedUserGettingRateLimited() throws InterruptedException {
+        authUser.setRateLimitTier(2); // 120 cals per hour - 1 added token every 30 seconds
+        boolean rateLimited = false;
+        int cnt;
+        for (cnt = 0; cnt <200; cnt++) {
+            rateLimited = !cache.checkRate(authUser, action);
+            if (rateLimited) {
+                break;
+            }
+        }
+        assertTrue(rateLimited && cnt == 120, "rateLimited:"+rateLimited + " cnt:"+cnt);
+
+        for (cnt = 0; cnt <60; cnt++) {
```

Comment:
**chore**: Please use Awaitility instead of such a loop with sleeps in it. See https://www.baeldung.com/java-thread-sleep-vs-awaitility-await for more examples.

Reply:
It's not waiting for an async action. It's waiting for a specific amount of time for the bucket to get more tokens.